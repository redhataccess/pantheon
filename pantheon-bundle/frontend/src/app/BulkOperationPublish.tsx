import React from 'react';

import "@app/app.css";
import { BulkPublishConfirmation } from './BulkPublishConfirmation';
import { BaseSizes, Button, Modal, ModalVariant, Title } from '@patternfly/react-core';
import { Utils } from './Utils';

export interface IBulkOperationPublishProps {
    documentsSelected: Array<{ cells: [string, { title: { props: { children: string[], href: string } } }, string, string, string], selected: boolean }>
    contentTypeSelected: string
    isBulkPublish: boolean
    isBulkUnpublish: boolean
    bulkOperationCompleted: boolean
    updateBulkOperationCompleted: (bulkOperationConfirmation) => any
}

class BulkOperationPublish extends React.Component<IBulkOperationPublishProps, any>{

    constructor(props) {
        super(props);
        this.state = {
            isModalOpen: true,

            showBulkConfirmation: false,

            //progress bar
            progressFailureValue: 0,
            progressSuccessValue: 0,
            progressWarningValue: 0,
            bulkUpdateFailure: 0,
            bulkUpdateSuccess: 0,
            bulkUpdateWarning: 0,

            docsSelected: this.props.documentsSelected,
            documentsSucceeded: [],
            documentsFailed: [""],
            documentsIgnored: [""],
            confirmationSucceeded: "",
            confirmationIgnored: "",
            confirmationFailed: "",
        };
    }

    public render() {
        const { isModalOpen } = this.state;
        const publishHeader = (
            <React.Fragment>
                <Title headingLevel="h1" size={BaseSizes["2xl"]}>
                    {this.props.isBulkPublish ? 'Publish' : 'Unpublish'}
            </Title>
            </React.Fragment>
        )
        const publishModal = (
            <React.Fragment>
                <Modal
                    variant={ModalVariant.medium}
                    title="Publish"
                    isOpen={this.state.isModalOpen}
                    header={publishHeader}
                    aria-label="Publish"
                    onClose={this.handleModalClose}
                    actions={[
                        <Button form="bulk_publish" key="publish" variant="primary" onClick={this.onBulkPublish}>
                            Publish
              </Button>,
                        <Button key="cancel" variant="secondary" onClick={this.handleModalClose}>
                            Cancel
                </Button>
                    ]}
                >
                    <div>
                        {this.props.contentTypeSelected == 'module' ? <div id="publish__module_helper_text"><p>Publishing <b>{this.props.documentsSelected.length}</b> module{this.props.documentsSelected.length > 1 ? 's.' : '.'}</p></div> : <div id="publish__assembly_helper_text"><p>Publishing <b>{this.props.documentsSelected.length}</b> assembl{this.props.documentsSelected.length > 1 ? 'ies.' : 'y.'}</p></div>}
                        <br />
                        <p>Publishing modules does <b>not</b> publish assemblies including these modules.</p>
                        <br />
                        <p>Publishing assemblies <b>does</b> publish all included draft modules. Modules removed from the published assembly are still published as separate modules.</p>
                        <br />
                        <p>If metadata is missing from any selected or included docs, they will <b>not</b> be published.</p>
                        <br />
                        <p>The publish process may take a while to update all files.</p>
                    </div>
                </Modal>
            </React.Fragment>
        );

        const unpublishModal = (
            <React.Fragment>
                <Modal
                    variant={ModalVariant.medium}
                    title="Unpublish"
                    isOpen={this.state.isModalOpen}
                    header={publishHeader}
                    aria-label="Unpublish"
                    onClose={this.handleModalClose}
                    actions={[
                        <Button form="bulk_unpublish" key="unpublish" variant="primary" onClick={this.onBulkPublish}>
                            Unpublish
              </Button>,
                        <Button key="cancel" variant="secondary" onClick={this.handleModalClose}>
                            Cancel
                </Button>
                    ]}
                >
                    <div>
                        {this.props.contentTypeSelected == 'module' ? <div id="unpublish__module_helper_text"><p>Unpublishing <b>{this.props.documentsSelected.length}</b> module{this.props.documentsSelected.length > 1 ? 's.' : '.'}</p></div> : <div id="unpublish__assembly_helper_text"><p>Unpublishing <b>{this.props.documentsSelected.length}</b> assembl{this.props.documentsSelected.length > 1 ? 'ies.' : 'y.'}</p></div>}
                        <br />
                        <p>Unpublishing assemblies does <b>not</b> unpublish included modules.</p>
                        <br />
                        <p>If selected modules are included in assemblies, they will <b>not</b> be unpublished.</p>
                    </div>
                </Modal>
            </React.Fragment>
        );

        return (
            <React.Fragment>
                {this.state.showBulkConfirmation &&
                    <BulkPublishConfirmation
                        key={new Date().getTime()}
                        header={this.props.isBulkPublish ? "Bulk Publish" : "Bulk Unpublish"}
                        subheading="Documents updated in the bulk operation"
                        updateSucceeded={this.state.confirmationSucceeded}
                        updateIgnored={this.state.confirmationIgnored}
                        updateFailed={this.state.confirmationFailed}
                        footer=""
                        progressSuccessValue={this.state.progressSuccessValue}
                        progressFailureValue={this.state.progressFailureValue}
                        progressWarningValue={this.state.progressWarningValue}
                        onShowBulkOperationConfirmation={this.updateShowBulkPublishConfirmation}
                        isBulkUnpublish={this.props.isBulkUnpublish}
                        bulkOperationCompleted={this.props.bulkOperationCompleted}
                        updateBulkOperationCompleted={this.props.updateBulkOperationCompleted}
                    />}

                {this.props.isBulkPublish && publishModal}
                {this.props.isBulkUnpublish && unpublishModal}
            </React.Fragment>
        );
    }

    private handleModalClose = () => {
        this.setState({ isModalOpen: false }, () => {
            this.props.updateBulkOperationCompleted(false)
        })
    }

    private onBulkPublish = (event) => {
        const formData = new FormData();

        {this.props.isBulkPublish ? formData.append(":operation", "pant:publish") : formData.append(":operation", "pant:unpublish")}


        const hdrs = {
            "Accept": "application/json",
            "cache-control": "no-cache",
            "Access-Control-Allow-Origin": "*",
        }
        let variant
        if(this.props.documentsSelected){
            variant = this.props.documentsSelected[0].cells[1].title.props.href.split("?variant=")[1]
            formData.append("variant", variant)
        }
        
        formData.append("locale", "en_US")

        // reinitialize states
        if (this.props.documentsSelected.length > 0) {
            this.setState({
                documentsSucceeded: [],
                documentsFailed: [""],
                documentsIgnored: [""],
                confirmationSucceeded: "",
                confirmationIgnored: "",
                confirmationFailed: "",
            });
        }

        this.props.documentsSelected.map((r) => {
            if (r.cells[1].title.props.href) {
                let href = r.cells[1].title.props.href

                //href part is module path
                let hrefPart = href.slice(0, href.indexOf("?"))
                let modulePath = hrefPart.slice(hrefPart.indexOf("/repositories"))
                const backend = "/content" + modulePath + `/en_US/variants/${variant}/draft`

                Utils.draftExist(backend).then((exist) => {
                    if (exist || this.props.isBulkUnpublish) {
                        fetch("/content" + modulePath, {
                            body: formData,
                            method: "post",
                            headers: hdrs
                        }).then(response => {
                            if (response.status === 201 || response.status === 200) {
                                console.log(this.props.isBulkPublish ? "publish works: " + response.status : "unpublish works: " + response.status)
                                this.setState({
                                    documentsSucceeded: [...this.state.documentsSucceeded, modulePath],
                                    bulkUpdateSuccess: this.state.bulkUpdateSuccess + 1,
                                }, () => {
                                    this.calculateSuccessProgress(this.state.bulkUpdateSuccess)
                                }
                                )
                            } else {
                                console.log(this.props.isBulkPublish ? "publish failed " + response.status : "unpublish failed " + response.status)
                                this.setState({ bulkUpdateFailure: this.state.bulkUpdateFailure + 1, documentsFailed: [...this.state.documentsFailed, modulePath] }, () => {
                                    this.calculateFailureProgress(this.state.bulkUpdateFailure)
                                })
                            }
                        })
                    } else {
                        console.log('no draft exists')
                        this.setState({ bulkUpdateWarning: this.state.bulkUpdateWarning + 1, documentsIgnored: [...this.state.documentsIgnored, modulePath] }, () => {
                            this.calculateWarningProgress(this.state.bulkUpdateWarning)
                        })
                    }
                })
            }

        })
        this.setState({ isModalOpen: false, showBulkConfirmation: true }, ()=>{
            this.props.updateBulkOperationCompleted(true)
        })
    }

    //functions for success & failure messages
    private updateShowBulkPublishConfirmation = (showBulkConfirmation) => {
        this.setState({ showBulkConfirmation })
    }

    private calculateFailureProgress = (num: number) => {
        if (num >= 0) {
            let stat = (num) / this.state.docsSelected.length * 100
            this.setState({ progressFailureValue: stat, showBulkConfirmation: true }, () => {
                this.getDocumentFailed()
            })
        }
    }

    private calculateSuccessProgress = (num: number) => {
        if (num >= 0) {
            let stat = (num) / this.state.docsSelected.length * 100
            this.setState({ progressSuccessValue: stat, showBulkConfirmation: true }, () => {
                this.getDocumentsSucceeded()
            })
        }
    }

    private calculateWarningProgress = (num: number) => {
        if (num >= 0) {
            let stat = (num) / this.state.docsSelected.length * 100
            this.setState({ progressWarningValue: stat, showBulkConfirmation: true }, () => {
                this.getDocumentIgnored()
            })
        }
    }
    private getDocumentsSucceeded = () => {
        if (this.state.documentsSucceeded.length > 0) {
            let succeeded = this.state.documentsSucceeded.join(",")
            this.setState({ confirmationSucceeded: succeeded })
        }
    }

    private getDocumentIgnored = () => {
        if (this.state.documentsIgnored.length > 0) {
            let ignored = this.state.documentsIgnored.join(",")
            this.setState({ confirmationIgnored: ignored })
        }
    }

    private getDocumentFailed = () => {
        if (this.state.documentsFailed.length > 0) {
            let failed = this.state.documentsFailed.join(",")
            this.setState({ confirmationFailed: failed })
        }
    }
}
export { BulkOperationPublish }