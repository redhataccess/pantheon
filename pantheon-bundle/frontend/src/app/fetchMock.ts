if (!self.fetch) {
    self.fetch = (input: RequestInfo, init?: RequestInit | undefined) => {
        const p = new Promise<Response>((resolve, reject) => {
            // just don't do anything. this exists so that our tests don't try and actually resolve fetches,
            // (which is exactly what isomorphic-fetch tried to do while we were importing it...)
            // and calling nothing here allows us to skip all of our fetch.then() and fetch.catch() code.
        })
        return p
    }
}