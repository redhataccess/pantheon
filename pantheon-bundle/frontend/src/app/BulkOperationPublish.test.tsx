import React from "react"
import { BulkOperationPublish } from "./BulkOperationPublish";

import { mount, shallow } from "enzyme"
import { Modal } from "@patternfly/react-core"
import { render } from '@testing-library/react'
import fetchMock from 'fetch-mock';

const anymatch = require("anymatch")
const props = {
    documentsSelected: anymatch,
    contentTypeSelected: "module",
    isBulkPublish: true,
    isBulkUnpublish: false,
    bulkOperationCompleted: true,
    updateIsBulkPublish: (isBulkPublish) => anymatch,
    updateIsBulkUnpublish: (isBulkUnpublish) => anymatch,
    updateBulkOperationCompleted: (bulkOperationCompleted) => anymatch
}

describe("BulkOperationPublish tests", () => {
    const api = "/content/products.harray.1.json"

    const response = {
        body: { __name__: "test product" },
        headers: { "content-type": "application/json" }
    }

    const Button = ({ onClick, children }) => (
        <button onClick={onClick}>{children}</button>
    )

    beforeEach(() => {
        fetchMock.getOnce(api, response)
    })


    afterEach(() => {
        fetchMock.reset();
        fetchMock.restore();
    });

    test("should render BulkOperationPublish component", () => {
        const view = shallow(<BulkOperationPublish {...props} />)
        expect(view).toMatchSnapshot()
    })

    it("should render a Modal component", () => {

        const wrapper = mount(<BulkOperationPublish {...props} />)
        const modal = wrapper.find(Modal)
        expect(modal.exists()).toBe(true)
    })

    it("test handleModalToggle function", () => {
        const { queryAllByText } = render(<BulkOperationPublish {...props} />)
        expect(queryAllByText("The publish process may take a while to update all files")).toBeTruthy()
        expect(queryAllByText("Publish")).toBeTruthy()
    })

})
