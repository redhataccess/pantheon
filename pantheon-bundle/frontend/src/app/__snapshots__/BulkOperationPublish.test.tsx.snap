// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkOperationPublish tests should render BulkOperationPublish component 1`] = `
<Fragment>
  <Modal
    actions={
      Array [
        <Button
          form="bulk_publish"
          onClick={[Function]}
          variant="primary"
        >
          Publish
        </Button>,
        <Button
          onClick={[Function]}
          variant="secondary"
        >
          Cancel
        </Button>,
      ]
    }
    appendTo={[Function]}
    aria-describedby=""
    aria-label="Publish"
    aria-labelledby=""
    className=""
    hasNoBodyWrapper={false}
    header={
      <React.Fragment>
        <Title
          headingLevel="h1"
          size="2xl"
        >
          Publish
        </Title>
      </React.Fragment>
    }
    isOpen={true}
    onClose={[Function]}
    ouiaSafe={true}
    showClose={true}
    title="Publish"
    titleIconVariant={null}
    titleLabel=""
    variant="medium"
  >
    <div>
      <div
        id="publish__module_helper_text"
      >
        <p>
          Publishing 
          <b>
            5
          </b>
           module
          s.
        </p>
      </div>
      <br />
      <p>
        Publishing modules does 
        <b>
          not
        </b>
         publish assemblies including these modules.
      </p>
      <br />
      <p>
        Publishing assemblies 
        <b>
          does
        </b>
         publish all included draft modules. Modules removed from the published assembly are still published as separate modules.
      </p>
      <br />
      <p>
        If metadata is missing from any selected or included docs, they will 
        <b>
          not
        </b>
         be published.
      </p>
      <br />
      <p>
        The publish process may take a while to update all files.
      </p>
    </div>
  </Modal>
</Fragment>
`;
